### PR TITLE
chore (pkg/babe): Integrate scale into babe and transaction libraries

### PR DESCRIPTION
**Pull request by jimjbrettj**
## Changes

- Integrated new scale pkg into lib/transactions
- Integrated scale into babe
- Refactored babe to utilize new scale features. This will be expanded upon to utilize results to hopefully clean it up more.

## Tests

```
go test ./lib/babe -v
go test ./lib/transaction -v

Note: There are errors thrown while testing with regards to block authoring (I don't think related to me) but wanted to check. Tests were still passing but wanted to get that checked out anyways.
```

## Issues

-  #1621

### CODE REVIEW

File: lib/babe/build.go [hunk 1]
```diff
@@ -347,16 +348,17 @@ func hasSlotEnded(slot Slot) bool {
 	return time.Since(slotEnd) >= 0
 }
 
-// ExtrinsicsToBody returns scale encoded block body which contains inherent and extrinsic.
+// ExtrinsicsToBody returns scale2 encoded block body which contains inherent and extrinsic.
```
Comment (noot) [inline lib/babe/build.go:351]: scale2?

File: lib/babe/errors.go [hunk 2]
```diff
@@ -81,73 +79,180 @@ func (e TransactionValidityError) Error() string {
...
+func (err Future) Index() uint {return 2}
+func (err Stale) Index() uint {return 3}
+func (err BadProof) Index() uint {return 4}
+func (err AncientBirthBlock) Index() uint {return 5}
+func (err ExhaustsResources) Index() uint {return 6}
+func (err InvalidCustom) Index() uint {return 7}
+func (err BadMandatory) Index() uint {return 8}
+func (err MandatoryDispatch) Index() uint {return 9}
+
+func determineDispatchErr(res []byte) error { // This works yay!
```
Comment (noot) [inline lib/babe/errors.go:175]: nice comment but pls remove :D

File: lib/babe/errors.go [hunk 3]
```diff
@@ -81,73 +79,180 @@ func (e TransactionValidityError) Error() string {
...
+	var e Other
+	vdt := scale.MustNewVaryingDataType(e, CannotLookup{}, BadOrigin{}, Module{})
+	err := scale.Unmarshal(res, &vdt)
 	if err != nil {
-		return err
+		return errInvalidResult
```
Comment (noot) [inline lib/babe/errors.go:180]: this masks the error coming from SCALE, might be nice to wrap it

File: lib/babe/build.go (hunk 1 above)
Comment (jimjbrettj) [reply lib/babe/build.go:351]: Had named it scale2 while integrating with the current scale library, forgot to remove from the comments will fix that.

File: lib/babe/errors.go (hunk 3 above)
Comment (jimjbrettj) [reply lib/babe/errors.go:180]: So would that be something like `return errors.New(<scale error>)`?

File: lib/babe/build.go [hunk 4]
```diff
@@ -347,16 +348,17 @@ func hasSlotEnded(slot Slot) bool {
...
+		var decExt []byte
+		err := scale.Unmarshal(tx.Extrinsic, &decExt)
 		if err != nil {
 			return nil, err
 		}
-		extrinsics = append(extrinsics, decExt.([]byte))
+		extrinsics = append(extrinsics, decExt)
 	}
 
 	return types.NewBodyFromExtrinsics(extrinsics)
```
Comment (timwu20) [inline lib/babe/build.go:364]: You should be able to unmarshal the whole `types.Body` using the scale package.  `NewBodyFromExtrinsics` is using the old package.

File: lib/babe/errors.go [hunk 5]
```diff
@@ -81,73 +79,180 @@ func (e TransactionValidityError) Error() string {
...
+}
+
+type Module struct { // add in `scale:"1"` after
+	Idx uint8
+	Err   uint8
+	Message *string
+}
+
+// Dispatch Receivers
+func (err Other) Index() uint {return 0}
```
Comment (timwu20) [inline lib/babe/errors.go:100]: nit: place the receiver functions below the type definition

File: lib/babe/errors.go [hunk 6]
```diff
@@ -81,73 +79,180 @@ func (e TransactionValidityError) Error() string {
...
 	}
+
 	return errInvalidResult
 }
 
+
 func determineErr(res []byte) error {
 	switch res[0] {
-	case 0: // DispatchOutcome
+	case 0:
```
Comment (timwu20) [inline lib/babe/errors.go:255]: will you be using `scale.Result` for this?

**Review by timwu20 — CHANGES_REQUESTED** [review]

File: lib/babe/build.go (hunk 4 above)
Comment (jimjbrettj) [reply lib/babe/build.go:364]: Okay I will look into doing this

File: lib/babe/errors.go [hunk 7]
```diff
@@ -79,105 +80,131 @@ func (e TransactionValidityError) Error() string {
 	return fmt.Sprintf("transaction validity error: %s", e.msg)
 }
 
-// Dispatch Errors
-type Other string
-
-type CannotLookup struct {
-	err string
+// A UnmarshallingError is when unmarshalling fails
+type UnmarshallingError struct {
```
Comment (timwu20) [inline lib/babe/errors.go:84]:
```suggestion
type UnmarshalError struct {
```

File: lib/babe/errors.go (hunk 6 above)
Comment (timwu20) [reply lib/babe/errors.go:281]: will this be done in this PR?

**Review by noot — APPROVED** [review]
nice work :D

File: lib/babe/errors.go (hunk 6 above)
Comment (jimjbrettj) [reply lib/babe/errors.go:281]: yes working on this now

**Review by EclesioMeloJunior — APPROVED** [review]

**Pull request closed without merging**